runtime/jcl/common/jclvm.c

```diff
@@ -138,6 +152,166 @@ Java_com_ibm_oti_vm_VM_allInstances(JNIEnv * env, jclass unused, jclass clazz, j
 	return count;
 }
 
+/**
+ * Return a String object containing a summary of the classes on the heap, 
+ * the number of instances, and their aggregate size.
+ * This string inserts Unix-style line separators.  The caller is responsible for translating them if necessary.
+ */
+jstring JNICALL
+Java_openj9_tools_attach_diagnostics_base_DiagnosticUtils_getHeapClassStatisticsImpl(JNIEnv * env, jclass unused)
+{
+	J9VMThread *vmThread = (J9VMThread *) env;
+	J9JavaVM *vm = vmThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9HashTable *statsTable = NULL;
+	J9HashTableState hashTableState;
+	BOOLEAN outOfMemory = FALSE;
+	j9object_t stringObject = NULL;
+	J9HeapStatisticsTableEntry **statsArray = NULL;
+	jstring stringObjectRef = NULL;
+	I_32 numClasses = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	vmFuncs->internalEnterVMFromJNI(vmThread);
+
+	vmFuncs->acquireExclusiveVMAccess(vmThread);
+	statsTable = collectHeapStatistics(vmThread);
+	vmFuncs->releaseExclusiveVMAccess(vmThread);
+
+	if (NULL != statsTable) {
+		numClasses = hashTableGetCount(statsTable);
+		statsArray = j9mem_allocate_memory(numClasses * sizeof(J9HeapStatisticsTableEntry*), J9MEM_CATEGORY_VM_JCL);
+	}
+	if (NULL == statsArray) {
+		outOfMemory = TRUE;
+	} else {
+		I_32 cursor = 0;
```

Comment:
No reason for this to be small or signed - why not UDATA?

Reply:
I agree it should be unsigned, but if we change it to `UDATA` it has implications for `printHeapStatistics()` and related variables like `I_32 numClasses`.

Reply:
Fixing the uses of 1`I_32`.

Reply:
I'll argue that using a signed type for something called `num` is patently incorrect. We should also (almost) never be using types smaller than UDATA. Using a 32-byte type for the count means limitting the number of instances on 64-bit unneccessarily. There's no overflow checking, so using properly-sized types seems important.